ambry-quota/src/main/java/com/github/ambry/quota/capacityunit/InMemoryCUQuotaSource.java

```diff
@@ -0,0 +1,205 @@
+  private final ScheduledExecutorService usageRefresher;
+  private final long aggregationWindowsInSecs;
+  private final AtomicBoolean isReady;
+  protected CapacityUnit feUsage; // Ambry frontend's CU usage.
+
+  /**
+   * Constructor for {@link InMemoryCUQuotaSource}.
+   * @param quotaConfig {@link QuotaConfig} object.
+   * @param accountService {@link AccountService} object.
+   * @throws IOException in case of any exception.
+   */
+  public InMemoryCUQuotaSource(QuotaConfig quotaConfig, AccountService accountService) throws IOException {
+    feQuota = JsonCUQuotaDataProviderUtil.getFeCUCapacityFromJson(quotaConfig.frontendCUCapacityInJson);
+    cuQuota = JsonCUQuotaDataProviderUtil.getCUQuotasFromJson(quotaConfig.resourceCUQuotaInJson, accountService);
+    cuUsage = new HashMap<>();
+    usageRefresher = Utils.newScheduler(1, REFRESHER_THREAD_NAME_PREFIX, true);
+    aggregationWindowsInSecs = quotaConfig.cuQuotaAggregationWindowInSecs;
+    isReady = new AtomicBoolean(false);
+  }
+
+  @Override
+  public synchronized void init() {
+    if (isReady.get()) {
+      LOGGER.warn("InMemoryCuQuotaSource is already initialized.");
+      return;
+    }
+    feUsage = new CapacityUnit();
+    cuQuota.keySet().forEach(key -> cuUsage.put(key, new CapacityUnit()));
+    usageRefresher.scheduleAtFixedRate(this::resetQuotaUsage, aggregationWindowsInSecs, aggregationWindowsInSecs,
+        TimeUnit.SECONDS);
+    isReady.set(true);
+  }
+
+  @Override
+  public boolean isReady() {
+    return isReady.get();
+  }
+
+  @Override
+  public Quota getQuota(QuotaResource quotaResource, QuotaName quotaName) throws QuotaException {
+    checkSupported(quotaName, quotaResource);
+
+    String resourceId = quotaResource.getResourceId();
+    if (cuQuota.containsKey(resourceId)) {
+      return new Quota<>(quotaName, cuQuota.get(resourceId).getQuotaValue(quotaName), quotaResource);
+    }
+    throw new QuotaException(String.format("Couldn't find quota for resource %s", quotaResource), true);
+  }
+
+  @Override
+  public float getUsage(QuotaResource quotaResource, QuotaName quotaName) throws QuotaException {
+    checkSupported(quotaName, quotaResource);
+    String resourceId = quotaResource.getResourceId();
+    if (!cuQuota.containsKey(resourceId)) {
+      throw new QuotaException(String.format("Couldn't find quota for resource %s", quotaResource), true);
+    }
+    double usage = 0;
+    if (cuUsage.containsKey(resourceId)) {
+      usage = cuUsage.get(resourceId).getQuotaValue(quotaName);
+    }
+    return QuotaUtils.getUsagePercentage(cuQuota.get(resourceId).getQuotaValue(quotaName), usage);
+  }
+
+  @Override
+  public void chargeUsage(QuotaResource quotaResource, QuotaName quotaName, double usageCost) throws QuotaException {
+    checkSupported(quotaName, quotaResource);
+    String resourceId = quotaResource.getResourceId();
+    if (!cuQuota.containsKey(resourceId)) {
```

Comment:
nit: This pattern repeats in many functions. You may consider creating a private function for this something like

private void assertResourceId(String resourceId) throws QuotaException {
}

Should we also emit a metric & create an alert for resource ids running without a quota?

Reply:
refactored the repeated pattern.
Will put out a separate PR for metrics. It possible that same metrics can be emitted at multiple levels (QuotaManager, QuotaSource, QuotaEnforcer or OperationController), and hence it will be easier to reason about metric placement once a good implementation is in shape.